### PR TITLE
[InstSimplify] Provide information about the range of possible values that `ucmp`/`scmp` can return

### PR DESCRIPTION
**Pull request by Poseydon42**
This makes it possible to fold dumb comparisons like `ucmp(x, y) == 7`.

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -9393,6 +9393,10 @@ static ConstantRange getRangeForIntrinsic(const IntrinsicInst &II) {
     if (!II.getParent() || !II.getFunction())
       break;
     return getVScaleRange(II.getFunction(), Width);
+  case Intrinsic::scmp:
+  case Intrinsic::ucmp:
+    return ConstantRange::getNonEmpty(APInt::getAllOnes(Width),
+                                      APInt::getOneBitSet(Width, 0));
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:9399]:
```suggestion
                                      APInt(Width, 2));
```
The upper bound is exclusive, so the range should be `[-1, 2)`, while you're currently constructing `[-1, 1)`.

**Review by nikic — COMMENTED** [review]
In terms of high level approach, an alternative would be to instead add the `range(-1, 2)` return attribute to the intrinsic. This has the advantage that all places that understand range attributes will automatically gain support for this.

The downside is that doing this is not entirely straightforward. One could easily add the range to callsites in InstCombine, but it would be better to generate it directly on the declaration. That would be done by adding a new intrinsic property to https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/Intrinsics.td and then generating a call to the appropriate constructor in https://github.com/llvm/llvm-project/blob/e7622ab4721141d9e6af6041fa7f9bbc1029e9aa/llvm/utils/TableGen/IntrinsicEmitter.cpp#L420. The tricky thing is that in this case, while the range itself is fixed, the bit width of the range depends on the type overload, which is not the case for any of the existing intrinsic properties. So I think `Intrinsic::getAttributes()` would need an extra `FunctionType *` argument, so that the attributes for the correct type could be emitted.

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 1 above)
Comment (Poseydon42) [reply llvm/lib/Analysis/ValueTracking.cpp]: Thanks, fixed.

**Comment by Poseydon42** [comment]
I don't know how many places do use the range attribute to aid their optimizations, but if that number is not too high then I feel like the current approach should be preferred to inventing something new for a single use case. Additionally, I think that in many places where range information is used, the actual range can be shrunk depending on the ranges of the arguments (e.g. the default `[-1,2)` range can be shrunk down to `[-1,1)` when we know that `LHS <= RHS`. Therefore, I think it may be more optimal to implement this sort of check in the places that do use the range info separately, instead of adding this information on the level of intrinsic itself.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by nikic** [comment]
The point of encoding this in the intrinsic attributes is that it's *not* for a specific use case, but a general mechanism. Target-independent intrinsics with fixed result ranges are pretty rare, but this comes up for target-specific intrinsics as well. For example arm_mve_pred_v2i currently adds a fixed range attribute via an InstCombine hook: https://github.com/llvm/llvm-project/blob/c19028f364cceb4b2111c7956dcacbc257a96fd4/llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp#L195C19-L195C35 x86_sse42_crc32_64_64 instead adds a special case in ValueTracking: https://github.com/llvm/llvm-project/blob/c19028f364cceb4b2111c7956dcacbc257a96fd4/llvm/lib/Analysis/ValueTracking.cpp#L1732C23-L1732C44 We have some NVPTX intrinsics with fixed ranges as well.

Anyway, we should still land this patch in the meantime. I may look into the generic mechanism myself.

**Comment by dtcxzyw** [comment]
Do you have a plan to fold some existing patterns into `ucmp/scmp`?

**Comment by Poseydon42** [comment]
I do, but probably after the main bulk of optimizations for them is done.

**Pull request merged**